src/topology/builder.rs

```diff
@@ -212,14 +220,29 @@ pub async fn build_pieces(
             schema_definitions.insert(output.port, schema_definition);
         }
 
+        let (pump_error_tx, mut pump_error_rx) = oneshot::channel();
         let pump = async move {
-            let mut handles = Vec::new();
+            debug!("Source pump supervisor starting.");
+
+            // Spawn all of the per-output pumps and then await their completion.
+            //
+            // If any of the pumps complete with an error, or panic/are cancelled, we return
+            // immediately.
+            let mut handles = FuturesUnordered::new();
             for pump in pumps {
                 handles.push(spawn_named(pump, task_name.as_ref()));
             }
-            for handle in handles {
-                handle.await.expect("join error")?;
+
+            while let Some(output) = handles.try_next().await? {
+                if let Err(e) = output {
+                    // Immediately send the error to the source's wrapper future, but ignore any
+                    // errors during the send, since nested errors wouldn't make any sense here.
+                    let _ = pump_error_tx.send(e);
+                    break;
+                }
             }
+
+            debug!("Source pump supervisor task finished normally.");
```

Comment:
Can we say that if we encountered an error above? That is, did it "finish normally" if it handled an error?

Reply:
Hmm, yeah, good catch.

Reply:
Fixed this in ddd91a8d6 to emit an appropriate message if one of the supervised pump tasks fails with an error.